File: _/aws-ch2-samples/src/com/packtpub/a3ws/ch2/samples/crud/CreateNodeWithContentService.java
package com.packtpub.a3ws.ch2.samples.crud;

import java.rmi.RemoteException;

import org.alfresco.webservice.content.Content;
import org.alfresco.webservice.repository.RepositoryFault;
import org.alfresco.webservice.repository.UpdateResult;
import org.alfresco.webservice.types.CML;
import org.alfresco.webservice.types.CMLCreate;
import org.alfresco.webservice.types.ContentFormat;
import org.alfresco.webservice.types.NamedValue;
import org.alfresco.webservice.types.ParentReference;
import org.alfresco.webservice.types.Predicate;
import org.alfresco.webservice.types.Reference;
import org.alfresco.webservice.types.Store;
import org.alfresco.webservice.util.AuthenticationUtils;
import org.alfresco.webservice.util.Constants;
import org.alfresco.webservice.util.ISO9075;
import org.alfresco.webservice.util.Utils;
import org.alfresco.webservice.util.WebServiceFactory;

public class CreateNodeWithContentService {

	/**
	 * @param args
	 * @throws RemoteException 
	 * @throws RepositoryFault 
	 */
	public static void main(String[] args) throws RepositoryFault, RemoteException {
		AuthenticationUtils.startSession("admin", "admin");
		Store spacesStore = new Store(Constants.WORKSPACE_STORE, "SpacesStore");
		String name = "AWS Book - Chapter 2 - "+System.currentTimeMillis();
		String description = "This is a content created with a sample of the book";
		String mimeType = "text/plain";
		String encoding = "UTF-8";
		
		try {

			ParentReference parent = new ParentReference(
					spacesStore, 
					null,
					"/app:company_home",
					Constants.ASSOC_CONTAINS,
					"{" + Constants.NAMESPACE_CONTENT_MODEL + "}"+name);
			
			//build properties
			NamedValue[] properties = buildCustomProperties(name,name,description);
			
			//create operation
			CMLCreate create = new CMLCreate();
			create.setId("1");
			create.setParent(parent);
			create.setType(Constants.TYPE_CONTENT);
			create.setProperty(properties);
			
			//create the node reference
			Reference reference = new Reference();
			reference.setStore(spacesStore);
			reference.setPath("/app:company_home/cm:"+ISO9075.encode(name));
			
			//create the predicate
			Predicate predicate = new Predicate();
			predicate.setNodes(new Reference[]{reference});
			
			//build the CML object
			CML cml = new CML();
	        cml.setCreate(new CMLCreate[]{create});
	        
	        //perform a CML update for the node
	        UpdateResult[] result = WebServiceFactory.getRepositoryService().update(cml);
	        
	        //get the new node reference
	        Reference referenceForContent = result[0].getDestination();
	        
	        //create content with ContentService
	        ContentFormat format = new ContentFormat(mimeType, encoding);
	        Content content = WebServiceFactory.getContentService().write(
	        		referenceForContent,
	        		Constants.PROP_CONTENT,  
	        		"This is the content for the new node".getBytes(), 
	        		format);

	        System.out.println("Node created: "+content.getNode().getPath());
	        
		} finally {
			AuthenticationUtils.endSession();
		}

	}

	private static NamedValue[] buildCustomProperties(String name, String title, String description) {
		NamedValue[] properties = new NamedValue[3];
		properties[0] = Utils.createNamedValue(Constants.PROP_NAME, name);
		properties[1] = Utils.createNamedValue(Constants.PROP_TITLE, title);
		properties[2] = Utils.createNamedValue(Constants.PROP_DESCRIPTION, description);
		return properties;
	}

}

File: _/1520_05_PreFinals_samples/src/Samples/Crud/CreateNodeWithContentService.cs
using System;
using System.Text;
using Alfresco;
using Alfresco.RepositoryWebService;
using NUnit.Framework;
using Samples.Vo;

namespace Samples.Crud
{
	[TestFixture]
	public class CreateNodeWithContentService
	{
		[TestCase]
		public void CanCreateNodesWithContent()
		{
			AuthenticationUtils.startSession("admin", "admin");
			Store spacesStore = new Store(StoreEnum.workspace, "SpacesStore");
			String name = "AWS Book - Chapter 2 - " + DateTime.Now.Ticks;
			String description = "This is a content created with a sample of the book";
			String mimeType = "text/plain";
			String encoding = "UTF-8";
			
			//custom value object
			CreateSampleVO createSampleVo = Builder.BuildCreateSampleVO(name, name, description);
			
			try {
	
				ParentReference parent = new ParentReference(
						spacesStore, 
						null,
						"/app:company_home",
						Constants.ASSOC_CONTAINS,
						"{" + Constants.NAMESPACE_CONTENT_MODEL + "}" + name
				);
				
				//build properties
				NamedValue[] properties = Builder.BuildCustomProperties(createSampleVo);
				
				//create operation
				CMLCreate create = new CMLCreate();
				create.id = "1";
				create.parent = parent;
				create.type = Constants.TYPE_CONTENT;
				create.property = properties;
				
				//create the node reference
				Reference reference = new Reference();
				reference.store = spacesStore;
				reference.path = "/app:company_home/cm:" + ISO9075.Encode(name);
				
				//create the predicate
				Predicate predicate = new Predicate();
				predicate.Items = new Reference[]{ reference };
				
				//build the CML object
				CML cml = new CML();
		        cml.create = new CMLCreate[]{ create };
		        
		        //perform a CML update for the node
		        UpdateResult[] result = WebServiceFactory.getRepositoryService().update(cml);
		        
		        //get the new node reference
		        Alfresco.ContentWebService.Reference referenceForContent = Alfresco.ContentWebService.Reference.From(result[0].destination);
		        
		        //create content with ContentService
		        Alfresco.ContentWebService.ContentFormat format = new Alfresco.ContentWebService.ContentFormat(mimeType, encoding);
		        Alfresco.ContentWebService.Content content = WebServiceFactory.getContentService().write(
		        		referenceForContent,
		        		Constants.PROP_CONTENT,  
		        		new ASCIIEncoding().GetBytes("This is the content for the new node"),
		        		format
		        );

		        String expectedPath = "/app:company_home/cm:AWS_x0020_Book_x0020_-_x0020_Chapter_x0020_2_x0020_-_x0020_";
		        Assert.IsTrue(content.node.path.StartsWith(expectedPath));
		        
			} finally {
				AuthenticationUtils.endSession();
			}
		}
	}
}
